### ISSUE TITLE
Implement monotonic timers

### ISSUE THREAD

**Issue by ronawho** [chapel-lang/chapel#7536]
Currently, we only support a single `CHPL_TIMER=generic`, which is implemented using `gettimeofday()`

gettimeofday has all sorts of problems. Most notably, it's not monotonic. Here's a bunch of links on why gettimeofday isn't great:

-  https://blog.habets.se/2010/09/gettimeofday-should-never-be-used-to-measure-time
-  http://stackoverflow.com/questions/88/is-gettimeofday-guaranteed-to-be-of-microsecond-resolution
- http://stackoverflow.com/questions/12392278/measure-time-in-linux-time-vs-clock-vs-getrusage-vs-clock-gettime-vs-gettimeof

and even a cursory google search will give you a lot more results.

Ideally we want to use `clock_gettime()` when possible. It's not implemented on OSX so we'll probably want to use the mach timer (though `clock_get_time()` might be another option.)

Implementing these different timers is easy, the only tricky part for us will be detecting when clock_gettime is available and whether we need to link in librt with it. Ideally we want an autoconf like check that looks for clock_gettime and detects if we need to throw `-lrt` but our current build infrastructure doesn't really support that.

Here's a good link on timers and what platforms support what:

- http://stackoverflow.com/questions/12392278/measure-time-in-linux-time-vs-clock-vs-getrusage-vs-clock-gettime-vs-gettimeof

For OSX I'm pretty sure we want to use the mach timer (this is what qthreads and gasnet use) but clock_get_time might be an option:

- https://gist.github.com/jbenet/1087739
- http://stackoverflow.com/questions/5167269/clock-gettime-alternative-in-mac-os-x

Note that we believe this is also the reason we're currently skipping sleepTimeUnits on OSX (#4158). We're sleeping using qthreads mach timer, but timing with our gettimeofday timer and they have different levels of precision.

**Comment by panzone** [comment]
Hi @ronawho , I'm interested in this issue.

macOS supports, since Sierra, clock_gettime(CLOCK_MONOTONIC), however I don't know right now if there are some differences between this and the timer used in qthread and gasnet. 

For the building, why not specify a new CHPL_TIMER? We default to generic, but on supported platform we permit something like CHPL_TIMER=monotonic and use clock_gettime. Am I missing something here?

**Comment by ronawho** [comment]
@panzone Oh cool, I didn't realize OSX finally started supporting `clock_gettime()`.

It should be straightforward to implement the timers themselves and users could manually opt-in to specific timers. The tricky part is updating [`util/chplenv/chpl_timers.py`](https://github.com/chapel-lang/chapel/blob/master/util/chplenv/chpl_timers.py) to auto-detect/ infer when `clock_gettime()` could be used.

For OSX is it enough to check OS versions? (are there any concerns with regard to the homebrew install? I would imagine homebrew bottles get built for each OS version, but I'm not sure)

For linux, how do we detect if `clock_gettime()` is available? jemalloc, qthreads, and probably gasnet are using autoconf checks, but our build system doesn't use autoconf/cmake/anything. Also note that on linux glibc < 2.17 will require `-lrt`

I had been imagining completely separate source directories for the different timers (runtime/src/timers/generic/, runtime/src/timers/gettime, runtime/src/timers/mach, etc), and then having chplenv choose the default, but another option might be to just keep single "generic" source file and use feature detection macros like `_POSIX_TIMERS` and/or `_POSIX_MONOTONIC_CLOCK`

Misc thought: Would we wan to expose `CLOCK_MONOTONIC_COARSE` and other things too?

**Comment by panzone** [comment]
Yes, for macOS is it enough to check the OS version. This shouldn't be a problem with the homebrew install, since I'm expecting to default to gettimeofday on older versions.

The only way I can think of is to check with dlopen/dlsym but it's far from perfect.

**Comment by nspark** [comment]
What is the rationale for making the timer selection a runtime configuration parameter (as I interpret it from the name `CHPL_TIMER`) vs. a compile-time parameter (like [the clocks in C++'s `std::chrono`](http://en.cppreference.com/w/cpp/chrono)?

**Comment by bradcray** [comment]
Hi @nspark: I'm reasonably certain the intention is to have it be a compile-time choice, similar to CHPL_COMM, CHPL_TASKS, CHPL_LAUNCHER, CHPL_ATOMICS, etc. are today.

**Comment by nspark** [comment]
@bradcray What I mean is that one can mix timer types in a single C++ source (trivial example):

```c++
#include <chrono>

int main(void)
{
  std::chrono::time_point<std::chrono::system_clock>
    system_time = std::chrono::system_clock::now();

  std::chrono::time_point<std::chrono::steady_clock>
    steady_time = std::chrono::steady_clock::now();

  return 0;
}
```

This is potentially useful because one would want a monotonic clock (i.e., `steady_clock`) for timing purposes, but only `system_clock` is convertible to a `time_t` type (e.g., for printing a timestamp with `strftime()`). Maybe this is why Chapel provides both DateTime and Time modules.

**Comment by ronawho** [comment]
@nspark That's a good point that we may want to support different/mixed timers in source code. Maybe open a separate issue for that?

For the "our timers aren't monotonic" part -- I think an easy/quick way to achieve this would be add support for something like`CHPL_TIMERS=qthreads`, and then if `CHPL_TASKS=qthreads` we could just default to that and we could use whatever timer qthreads uses. On most unix platforms this will be clock_gettime, and on macs it will be mach. 

This gets us monotonic timers in most places we care about without having to worry about detecting what timer works (and what links flags are needed) because qthreads is already doing that for us during its configure checks.

**Comment by mppf** [comment]
> it's not implemented on OSX so we'll probably want to use the mach timer 

My Mac OS X system shows a man page for `clock_gettime`. IIRC on Mac OS X you don't have to pass `-lrt` to get it but on linux you do.

**Comment by mppf** [comment]
One of the S.O. posts says it was added in 10.12 Sierra.

**Comment by mppf** [comment]
And now I see this was already discussed in earlier comments...

**Comment by mppf** [comment]
> For OSX is it enough to check OS versions? (are there any concerns with regard to the homebrew install? I would imagine homebrew bottles get built for each OS version, but I'm not sure)

Homebrew currently requires Mac OS X Mojave 10.14 or higher. So any Chapel runtime built for Homebrew can assume that `clock_gettime` exists.

I think it would be fine to use the "lean on qthreads" strategy or to use `clock_gettime` and add `-lrt` on glibc/linux systems.

**Comment by nspark** [comment]
FWIW, `-lrt` is only required for `clock_gettime` on Linux with glibc < 2.17 (see [release notes](https://sourceware.org/legacy-ml/libc-announce/2012/msg00001.html)):

```
 * The `clock_*' suite of functions (declared in <time.h>) is now available
  directly in the main C library.  Previously it was necessary to link with
  -lrt to use these functions.  This change has the effect that a
  single-threaded program that uses a function such as `clock_gettime' (and
  is not linked with -lrt) will no longer implicitly load the pthreads
  library at runtime and so will not suffer the overheads associated with
  multi-thread support in other code such as the C++ runtime library.
```
